### PR TITLE
cpu/cortex-m: turn MPU support into a feature

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

This turns the MPU of Cortex-M into a feature.

### CODE REVIEW

File: cpu/cortexm_common/Makefile.features [hunk 1]
```diff
@@ -5,3 +5,7 @@ FEATURES_PROVIDED += periph_pm
 FEATURES_PROVIDED += cpp
 FEATURES_PROVIDED += cpu_check_address
 FEATURES_PROVIDED += ssp
+
+ifneq ($(CPU_ARCH),cortex-m0)
+  FEATURES_PROVIDED += cortexm_mpu
+endif
```
Comment (benpicco) [inline cpu/cortexm_common/Makefile.features]: According to [Wikipedia](https://en.wikipedia.org/wiki/ARM_Cortex-M#Silicon_customization), the MPU is always optional.

Comment (kaspar030) [reply cpu/cortexm_common/Makefile.features]: Hmok. Do all our non-cm0 CPUs have an MCU?
I'll add a check to the mpu test application.

Comment (kaspar030) [reply cpu/cortexm_common/Makefile.features]: I've added a check. Seems like all our non-Cortex-M0 boards have an MPU, or at leastm `__MPU_PRESENT` set.

File: tests/mpu_stack_guard/Makefile.ci [hunk 2]
```diff
@@ -1,3 +1,4 @@
 BOARD_INSUFFICIENT_MEMORY := \
+    nucleo-f031k6 \
```
Comment (aabadie) [inline tests/mpu_stack_guard/Makefile.ci]: This board (and stm32f030f4-demo) are already excluded by the `FEATURES_REQUIRED += cortexm_mpu` and should already be marked as unsupported by the CI. Why do you add it here ?

Comment (kaspar030) [reply tests/mpu_stack_guard/Makefile.ci]: Hm, seems like the feature guard doesn't work:
```
[kaspar@ng riot.tmp/tests/mpu_stack_guard (mpu_feature)]$ make info-boards-supported | grep -o nucleo-f031k6
nucleo-f031k6
[kaspar@ng riot.tmp/tests/mpu_stack_guard (mpu_feature)]$ 
```

```
[kaspar@ng riot.tmp/tests/mpu_stack_guard (mpu_feature)]$ BOARD=nucleo-f031k6 make info-debug-variable-CPU_ARCH
cortex-m0
[kaspar@ng riot.tmp/tests/mpu_stack_guard (mpu_feature)]$
```

Seems like CPU_ARCH is not available in Makefile.features.

Comment (aabadie) [reply tests/mpu_stack_guard/Makefile.ci]: It seems that this PR tries to achieve the same thing as #13236 where CPU_ARCH is moved to Makefile.features. I can cut this part out of #13236 if it also helps here.

File: cpu/cortexm_common/Makefile.features (hunk 1 above)
Comment (kaspar030) [reply cpu/cortexm_common/Makefile.features]: Either `__MPU_PRESENT` is not undefined, or *all* our Cortex-M boards *do* have the optional MPU, even the CM0. @aabadie Could you verify the test succeeds e.g., on a stm32f031k6?

File: tests/mpu_stack_guard/Makefile.ci (hunk 2 above)
Comment (kaspar030) [reply tests/mpu_stack_guard/Makefile.ci]: > I can cut this part out of #13236 if it also helps here.

That would be awesome.

Comment (aabadie) [reply tests/mpu_stack_guard/Makefile.ci]: Here it is: #13392 :)

**Comment by benpicco** [comment]
So this needs a rebase now.

File: tests/mpu_stack_guard/Makefile.ci (hunk 2 above)
Comment (kaspar030) [reply tests/mpu_stack_guard/Makefile.ci]: Thanks a bunch.

```
[kaspar@ng riot.tmp/tests/mpu_stack_guard (mpu_feature)]$ make info-boards-supported | grep -o nucleo-f031k6
[kaspar@ng riot.tmp/tests/mpu_stack_guard (mpu_feature)]$
```

**Comment by kaspar030** [comment]
> So this needs a rebase now.

Done. Removed an unrelated ssp change, and renamed the WIP commit.

**Comment by kaspar030** [comment]
- fixed long commit msg

File: cpu/cortexm_common/Makefile.features (hunk 1 above)
Comment (kaspar030) [reply cpu/cortexm_common/Makefile.features]: @aabadie could you take a look at this? I don't have a M0 board...

**Comment by kaspar030** [comment]
Pausing for https://github.com/RIOT-OS/RIOT/pull/13456

File: cpu/cortexm_common/Makefile.features (hunk 1 above)
Comment (aabadie) [reply cpu/cortexm_common/Makefile.features]: I tried on nucleo-f072rb, which is an M0, and the `tests/mpu_stack_guard` is correctly skipped by the build system.

The problem here is that, according to the __MPU_PRESENT macro in vendor files, stm32f1 and also some stm32f3 (so not all) CPUs  doesn't provide an MPU. It's also unclear if stm32l0 provides an MPU or not (I'm wondering if the vendor files were manually altered for this family).

So, as suggested by @benpicco, I think this feature should be provided by default less widely. Maybe just for M4/M7 by default, and enable on a case-by-case basis for other CPU families (M0/M0plus,M3).

**Comment by kaspar030** [comment]
> So, as suggested by @benpicco, I think this feature should be provided by default less widely. Maybe just for M4/M7 by default, and enable on a case-by-case basis for other CPU families (M0/M0plus,M3).

Even for M4 this seems to not not always be available (e.g., stm32f302r8 is an M4, but doesn't set __MPU_PRESENT). Seaveing through all CPU's is quite a hassle.

I'm inclined to define this per-board, as that can be done quickly (by, e.g., use the buildtest output of all cortex-m, then add the feature to all board Makefile.features where it succeeded).
We can then move it to more common places later on.

@aabadie what do you think?

**Comment by aabadie** [comment]
> what do you think?

I agree that the feature should be provided more finely. Maybe provide the feature in whitelisted board as a first step ? (and no need to wait for buildtest ;))

**Comment by benpicco** [comment]
> Seaveing through all CPU's is quite a hassle.

We can just add it to known-good CPU families for now, it's easy to fix omissions later.
E.g. I think it's save to assume no stm32f3 has a MPU.

> I'm inclined to define this per-board

But this is not a board feature. We should get rid of MCU features in boards, not add more of those.

**Comment by kaspar030** [comment]
> E.g. I think it's save to assume no stm32f3 has a MPU.

Hmm,

> (e.g., stm32f302r8 is an M4, but doesn't set __MPU_PRESENT).

**Comment by kaspar030** [comment]
> But this is not a board feature. We should get rid of MCU features in boards, not add more of those.

Agreed, but the way our build configuration is modeled, there's no way to print the tree of dependencies (board -> cpu -> cpu_fam -> cpu_arch), and there are many layers of includes with non-standard names.

I can do the manual equivalent of `make buildtest | grep succeeded | cut -d" " -f3` to get a list of *boards* that work, and with that I can do `for board in $BOARDS; do echo "FEATURES_PROVIDED += cortexm_mpu" >> boards/$BOARD/Makefile.features`, then cleanup the result.

Doing it "the right way" will take maybe a day? We'd have to go through all ~180 board Makefiles, figure out which CPU family they're in, try to find commonalities, build, refine.

**Comment by benpicco** [comment]
> > E.g. I think it's save to assume *no* stm32f3 has a MPU.
> 
> Hmm,
> 
> > (e.g., stm32f302r8 is an M4, but doesn't set __MPU_PRESENT).

I see no contradiction 😉 

> Doing it "the right way" will take maybe a day? We'd have to go through all ~180 board Makefiles, figure out which CPU family they're in, try to find commonalities, build, refine.

I'm not advocating for a complete list, just adding `FEATURES_PROVIDED += cortexm_mpu` to the hand full of families where we know it's present.
Sure we'll miss some, but if people want to use the MPU on these boards they can add the `FEATURES_PROVIDED += cortexm_mpu`.

**Comment by kaspar030** [comment]
> I'm not advocating for a complete list, just adding `FEATURES_PROVIDED += cortexm_mpu` to the hand full of families where we know it's present.

Makes sense. Which ones are those?

**Comment by benpicco** [comment]
> Which ones are those?

stm32f4 (checked the [smallest member](https://www.st.com/resource/en/datasheet/stm32f401cb.pdf) of the family), nrf52, samd5x

File: cpu/cortexm_common/Makefile.features (hunk 1 above)
Comment (benpicco) [reply cpu/cortexm_common/Makefile.features]: I just checked the data sheets, neither samd2x nor saml2x (cm0p) implement an MPU.

**Comment by kaspar030** [comment]
> stm32f4 (checked the [smallest member](https://www.st.com/resource/en/datasheet/stm32f401cb.pdf) of the family), stm32l4 ([dito](https://www.st.com/resource/en/datasheet/stm32l471qe.pdf)), nrf51, nrf52, samd5x, cc26x2_cc13x2, lm4f120, sam3, saml1x, cc2538

Done. Removed nrf51, they don't have an MPU.

**Comment by benpicco** [comment]
> Removed nrf51, they don't have an MPU.

The [Data Sheet](https://infocenter.nordicsemi.com/pdf/nRF51824_PS_v1.0.pdf) has a chapter about an MPU though.

**Comment by kaspar030** [comment]
> The [Data Sheet](https://infocenter.nordicsemi.com/pdf/nRF51824_PS_v1.0.pdf) has a chapter about an MPU though.

Hm, maybe it is non-standard. The CMSIS headers define `__MPU_PRESENT` to 0, and if forced to 1, compilation fails, e.g.,:

```
/home/kaspar/src/riot/cpu/cortexm_common/mpu.c: In function 'mpu_disable':
/home/kaspar/src/riot/cpu/cortexm_common/mpu.c:28:5: error: 'MPU' undeclared (first use in this function); did you mean 'MPU_H'?
   28 |     MPU->CTRL &= ~MPU_CTRL_ENABLE_Msk;
      |     ^~~                                                                                                                  
      |     MPU_H                                               
```

Seems like "MPU" is what Nordic calls flash protection:

```
The memory protection unit can be configured to protect all flash memory on the device from readback, or to protect blocks of flash from over-write or erase.
```

**Comment by kaspar030** [comment]
> Could `FEATURES_PROVIDED += cortexm_mpu` be also added to `stm32f1`, `stm32f2` and `stm32l1` ?
> (initially iotlab-*, nucleo-l152re and nucleo-207zg were whitelisted in the test).

Added the stm32f2 and l1. The others don't define __MPU_PRESENT.

File: tests/mpu_stack_guard/main.c [hunk 3]
```diff
@@ -22,6 +22,14 @@
 
 #include "cpu.h"
 #include "thread.h"
+#include "mpu.h"
+
+/* RIOT's MPU headers gracefully fail when no MPU is present.
+ * Use this to catch if RIOT's features are correctly gateing MPU use.
```
Comment (aabadie) [inline tests/mpu_stack_guard/main.c]: gateing => gating ?

**Comment by aabadie** [comment]
> Added the stm32f2 and l1

Did you forget to push ?

File: tests/mpu_stack_guard/main.c (hunk 3 above)
Comment (kaspar030) [reply tests/mpu_stack_guard/main.c]: fixed

**Comment by kaspar030** [comment]
> Did you forget to push ?

yup :)

**Comment by kaspar030** [comment]
There are some missing that where whitelisted before and actually work. Will add them.

**Comment by kaspar030** [comment]
Now I got them all. Special cases were: _some_ stm32f303 and _some_ stm32l052 have an MPU.
The PR now covers all CPU's in master.
It was not so bad, thanks for pushing @benpicco @aabadie.

**Pull request merged**